File: test/ignore-paths.test.ts
/**
 * test/ignore-paths.test.ts — the `ignorePaths` capability (GEN §10, CONF §5.6).
 *
 * Covers the construction-time validation errors (GEN §10.1/GEN §10.7 — not
 * expressible as diff/apply vectors, CONF §6.2) plus the happy-path and
 * interaction semantics in both directions.
 */
import { describe, expect, test } from "bun:test";
import { buildPlan, JsonSchemaPatcher } from "../src/index";
import type { JsonValue, Operation } from "../src/types";

const EMPTY = () => new Map();

function diff(
  opts: {
    plan?: Map<string, never>;
    ignorePaths?: string[];
    emitMoves?: boolean;
    wholesaleReplaceFallback?: boolean;
    includeOldValue?: boolean;
    // biome-ignore lint: test convenience
    planObj?: any;
  },
  original: JsonValue,
  modified: JsonValue,
): Operation[] {
  const plan = opts.planObj
    ? buildPlan({ schema: opts.planObj })
    : (opts.plan ?? EMPTY());
  const p = new JsonSchemaPatcher({
    plan,
    ignorePaths: opts.ignorePaths,
    emitMoves: opts.emitMoves,
    wholesaleReplaceFallback: opts.wholesaleReplaceFallback,
    includeOldValue: opts.includeOldValue,
  });
  return p.execute({ original, modified });
}

const PK_USERS = {
  type: "object",
  properties: {
    users: {
      type: "array",
      items: {
        type: "object",
        properties: { id: { type: "string" }, updatedAt: { type: "number" }, name: { type: "string" } },
        required: ["id"],
      },
    },
  },
};

describe("ignorePaths validation (GEN §10.1)", () => {
  test("rejects a canonical array-index segment", () => {
    expect(() => new JsonSchemaPatcher({ plan: EMPTY(), ignorePaths: ["/users/0/id"] })).toThrow(
      TypeError,
    );
  });
  test("rejects the '-' segment", () => {
    expect(() => new JsonSchemaPatcher({ plan: EMPTY(), ignorePaths: ["/a/-"] })).toThrow(TypeError);
  });
  test("rejects the empty/root pointer", () => {
    expect(() => new JsonSchemaPatcher({ plan: EMPTY(), ignorePaths: [""] })).toThrow(TypeError);
  });
  test("rejects a pointer without a leading slash", () => {
    expect(() => new JsonSchemaPatcher({ plan: EMPTY(), ignorePaths: ["a/b"] })).toThrow(TypeError);
  });
  test("accepts a non-canonical numeric-looking segment (leading zero, object member)", () => {
    // "01" is a legal member name, not an array index (§F33).
    expect(() => new JsonSchemaPatcher({ plan: EMPTY(), ignorePaths: ["/x/01"] })).not.toThrow();
  });
  test("accepts '*' wildcard and normal members", () => {
    expect(
      () => new JsonSchemaPatcher({ plan: EMPTY(), ignorePaths: ["/users/*/updatedAt", "/meta/ts"] }),
    ).not.toThrow();
  });
  test("accepts an escaped member name containing a slash", () => {
    expect(() => new JsonSchemaPatcher({ plan: EMPTY(), ignorePaths: ["/a~1b/c"] })).not.toThrow();
  });
});

describe("ignorePaths primaryKey guard (GEN §10.7)", () => {
  test("rejects ignoring the primaryKey field itself", () => {
    const plan = buildPlan({ schema: PK_USERS });
    expect(() => new JsonSchemaPatcher({ plan, ignorePaths: ["/users/*/id"] })).toThrow(TypeError);
  });
  test("rejects ignoring the whole keyed item (covers the key)", () => {
    const plan = buildPlan({ schema: PK_USERS });
    expect(() => new JsonSchemaPatcher({ plan, ignorePaths: ["/users/*"] })).toThrow(TypeError);
  });
  test("rejects ignoring the whole keyed array (covers the key)", () => {
    const plan = buildPlan({ schema: PK_USERS });
    expect(() => new JsonSchemaPatcher({ plan, ignorePaths: ["/users"] })).toThrow(TypeError);
  });
  test("allows ignoring a NON-key field of a keyed item", () => {
    const plan = buildPlan({ schema: PK_USERS });
    expect(() => new JsonSchemaPatcher({ plan, ignorePaths: ["/users/*/updatedAt"] })).not.toThrow();
  });
  test("catches a primaryKey under an additionalProperties (*) plan path", () => {
    const schema = {
      type: "object",
      additionalProperties: {
        type: "array",
        items: {
          type: "object",
          properties: { id: { type: "string" } },
          required: ["id"],
        },
      },
    };
    const plan = buildPlan({ schema }); // plan key "/*"
    // ignore "/envA/*/id" covers the key field of /envA (matched via /* at diff time)
    expect(() => new JsonSchemaPatcher({ plan, ignorePaths: ["/envA/*/id"] })).toThrow(TypeError);
  });
});

describe("ignorePaths happy paths", () => {
  test("object member: ignored-only change -> []", () => {
    expect(diff({ ignorePaths: ["/meta/ts"] }, { meta: { ts: 1, n: "a" } }, { meta: { ts: 2, n: "a" } })).toEqual(
      [],
    );
  });
  test("object member: real change survives, ignored suppressed", () => {
    expect(
      diff({ ignorePaths: ["/meta/ts"] }, { meta: { ts: 1, n: "a" } }, { meta: { ts: 2, n: "b" } }),
    ).toEqual([{ op: "replace", path: "/meta/n", value: "b", oldValue: "a" }]);
  });
  test("wildcard: ignore any member at a level", () => {
    expect(diff({ ignorePaths: ["/*/ts"] }, { a: { ts: 1 }, b: { ts: 1, k: 2 } }, { a: { ts: 9 }, b: { ts: 9, k: 3 } })).toEqual(
      [{ op: "replace", path: "/b/k", value: 3, oldValue: 2 }],
    );
  });
  test("modified-only ignored member is not added", () => {
    expect(diff({ ignorePaths: ["/meta/ts"] }, { meta: {} }, { meta: { ts: 5 } })).toEqual([]);
  });
  test("whole array ignored via `/arr` -> []", () => {
    expect(diff({ ignorePaths: ["/arr"] }, { arr: [1, 2, 3] }, { arr: [4, 5] })).toEqual([]);
  });
  test("every element ignored via `/arr/*` -> []", () => {
    expect(diff({ ignorePaths: ["/arr/*"] }, { arr: [{ a: 1 }] }, { arr: [{ a: 2 }, { b: 3 }] })).toEqual([]);
  });
});

describe("ignorePaths under keyed arrays (GEN §10.3)", () => {
  test("primaryKey: ignored-field-only change -> []", () => {
    const ops = diff(
      { planObj: PK_USERS, ignorePaths: ["/users/*/updatedAt"] },
      { users: [{ id: "a", updatedAt: 1 }, { id: "b", updatedAt: 1 }] },
      { users: [{ id: "a", updatedAt: 2 }, { id: "b", updatedAt: 1 }] },
    );
    expect(ops).toEqual([]);
  });
  test("primaryKey: real field change survives while ignored drifts", () => {
    const ops = diff(
      { planObj: PK_USERS, ignorePaths: ["/users/*/updatedAt"] },
      { users: [{ id: "a", updatedAt: 1, name: "A" }] },
      { users: [{ id: "a", updatedAt: 2, name: "A2" }] },
    );
    expect(ops).toEqual([{ op: "replace", path: "/users/0/name", value: "A2", oldValue: "A" }]);
  });
});

describe("ignorePaths under LCS (GEN §10.5)", () => {
  test("ignored-field-only change -> []", () => {
    expect(
      diff({ ignorePaths: ["/items/*/ts"] }, { items: [{ v: 1, ts: 1 }, { v: 2, ts: 1 }] }, { items: [{ v: 1, ts: 9 }, { v: 2, ts: 1 }] }),
    ).toEqual([]);
  });
  test("move-pairing with ignored drift (emitMoves): a relocation, not remove+add", () => {
    const ops = diff(
      { emitMoves: true, ignorePaths: ["/items/*/ts"] },
      { items: [{ v: "X", ts: 1 }, { v: "Y", ts: 1 }] },
      { items: [{ v: "Y", ts: 9 }, { v: "X", ts: 1 }] },
    );
    // Y (ts drift) and X reorder -> exactly one move; NO replace for the ts drift.
    expect(ops.filter((o) => o.op === "replace")).toEqual([]);
    expect(ops.filter((o) => o.op === "move").length).toBe(1);
  });
});

describe("ignorePaths x wholesaleReplaceFallback (GEN §10.6)", () => {
  const RW = {
    type: "object",
    properties: {
      a: { type: "array", items: { type: "object", properties: { t: { type: "string" }, ts: { type: "number" } } } },
    },
  };
  test("wholesale disabled when an ignore path lies beneath the array", () => {
    const ops = diff(
      { planObj: RW, wholesaleReplaceFallback: true, ignorePaths: ["/a/*/ts"] },
      { a: [{ t: "AAAAAAAAAA", ts: 1 }, { t: "BBBBBBBBBB", ts: 2 }] },
      { a: [{ t: "ZZZZZZZZZZ", ts: 5 }, { t: "YYYYYYYYYY", ts: 6 }] },
    );
    // A whole-array replace at /a would carry the ignored `ts` values and leak
    // them; the capability is disabled, so we get a granular (ignore-filtered)
    // stream instead — no single replace whose path is exactly "/a".
    expect(ops.some((o) => o.op === "replace" && o.path === "/a")).toBe(false);
    // And no emitted op mentions ts (all touched fields are /t or whole items).
    const changed = ops.filter((o) => o.op === "replace");
    for (const o of changed) expect(o.path.endsWith("/ts")).toBe(false);
  });
});

describe("ignorePaths byte-stability (GEN §10.2)", () => {
  test("absent ignorePaths is byte-identical to the pre-capability output", () => {
    const orig = { meta: { ts: 1, n: "a" }, arr: [{ id: "x", v: 1 }] };
    const mod = { meta: { ts: 2, n: "b" }, arr: [{ id: "x", v: 2 }] };
    const base = diff({}, orig, mod);
    const withEmptyList = diff({ ignorePaths: [] }, orig, mod);
    expect(withEmptyList).toEqual(base);
  });
});

File: go/ignore_paths_test.go
package schemapatch

import (
	"strings"
	"testing"
)

// ignore_paths_test.go covers the ignorePaths capability (GEN §10, CONF §5.6),
// including the construction-time validation errors (GEN §10.1/GEN §10.7) that the
// diff vector wire format cannot express (CONF §6.2), plus happy-path and
// interaction semantics — mirroring test/ignore-paths.test.ts.

func ignoreDiff(t *testing.T, schema string, ignore []string, opts []PatcherOption, orig, mod string) []Operation {
	t.Helper()
	var plan Plan
	if schema != "" {
		var err error
		plan, err = BuildPlan(mustDecode(t, schema), BuildPlanOptions{})
		if err != nil {
			t.Fatalf("BuildPlan: %v", err)
		}
	}
	allOpts := append([]PatcherOption{IgnorePaths(ignore...)}, opts...)
	p, err := NewPatcher(plan, allOpts...)
	if err != nil {
		t.Fatalf("NewPatcher: %v", err)
	}
	return p.Execute(mustDecode(t, orig), mustDecode(t, mod))
}

const pkUsersSchema = `{"type":"object","properties":{"users":{"type":"array","items":{"type":"object","required":["id"],"properties":{"id":{"type":"string"},"updatedAt":{"type":"number"},"name":{"type":"string"}}}}}}`

func TestIgnorePathsValidation(t *testing.T) {
	cases := []struct {
		name   string
		schema string
		paths  []string
	}{
		{"array-index-segment", "", []string{"/users/0/id"}},
		{"dash-segment", "", []string{"/a/-"}},
		{"empty-root-pointer", "", []string{""}},
		{"no-leading-slash", "", []string{"a/b"}},
		{"pk-field-itself", pkUsersSchema, []string{"/users/*/id"}},
		{"whole-keyed-item", pkUsersSchema, []string{"/users/*"}},
		{"whole-keyed-array", pkUsersSchema, []string{"/users"}},
	}
	for _, tc := range cases {
		t.Run(tc.name, func(t *testing.T) {
			var plan Plan
			if tc.schema != "" {
				var err error
				plan, err = BuildPlan(mustDecode(t, tc.schema), BuildPlanOptions{})
				if err != nil {
					t.Fatalf("BuildPlan: %v", err)
				}
			}
			_, err := NewPatcher(plan, IgnorePaths(tc.paths...))
			if err == nil {
				t.Fatalf("expected NewPatcher to reject ignorePaths %v", tc.paths)
			}
			if !strings.Contains(err.Error(), "ignorePaths") {
				t.Errorf("error should mention ignorePaths, got %q", err.Error())
			}
		})
	}
}

func TestIgnorePathsAccepts(t *testing.T) {
	cases := [][]string{
		{"/x/01"},              // leading-zero segment is a legal member name, not an index
		{"/users/*/updatedAt"}, // wildcard + member
		{"/meta/ts"},
		{"/a~1b/c"}, // escaped member name containing a slash
	}
	for _, paths := range cases {
		if _, err := NewPatcher(Plan{}, IgnorePaths(paths...)); err != nil {
			t.Errorf("ignorePaths %v should be accepted, got %v", paths, err)
		}
	}
}

func TestIgnorePathsPrimaryKeyUnderAdditionalProps(t *testing.T) {
	// A primaryKey under an additionalProperties (*) plan path must still be
	// caught (GEN §10.7): plan key "/*", ignore "/envA/*/id".
	schema := `{"type":"object","additionalProperties":{"type":"array","items":{"type":"object","required":["id"],"properties":{"id":{"type":"string"}}}}}`
	plan, err := BuildPlan(mustDecode(t, schema), BuildPlanOptions{})
	if err != nil {
		t.Fatalf("BuildPlan: %v", err)
	}
	if _, err := NewPatcher(plan, IgnorePaths("/envA/*/id")); err == nil {
		t.Fatal("expected rejection: ignore covers the primaryKey field of the /* plan")
	}
}

func TestIgnorePathsHappyPaths(t *testing.T) {
	t.Run("obj-member-ignored-only", func(t *testing.T) {
		ops := ignoreDiff(t, "", []string{"/meta/ts"}, nil, `{"meta":{"ts":1,"n":"a"}}`, `{"meta":{"ts":2,"n":"a"}}`)
		if len(ops) != 0 {
			t.Fatalf("want [], got %s", dumpOps(ops))
		}
	})
	t.Run("obj-member-real-change-survives", func(t *testing.T) {
		ops := ignoreDiff(t, "", []string{"/meta/ts"}, nil, `{"meta":{"ts":1,"n":"a"}}`, `{"meta":{"ts":2,"n":"b"}}`)
		if len(ops) != 1 || string(ops[0].Op) != "replace" || ops[0].Path != "/meta/n" {
			t.Fatalf("want single replace /meta/n, got %s", dumpOps(ops))
		}
	})
	t.Run("whole-array-ignored", func(t *testing.T) {
		ops := ignoreDiff(t, "", []string{"/arr"}, nil, `{"arr":[1,2,3]}`, `{"arr":[4,5]}`)
		if len(ops) != 0 {
			t.Fatalf("want [], got %s", dumpOps(ops))
		}
	})
	t.Run("every-element-ignored", func(t *testing.T) {
		ops := ignoreDiff(t, "", []string{"/arr/*"}, nil, `{"arr":[{"a":1}]}`, `{"arr":[{"a":2},{"b":3}]}`)
		if len(ops) != 0 {
			t.Fatalf("want [], got %s", dumpOps(ops))
		}
	})
}

func TestIgnorePathsKeyedArray(t *testing.T) {
	t.Run("ignored-field-only", func(t *testing.T) {
		ops := ignoreDiff(t, pkUsersSchema, []string{"/users/*/updatedAt"}, nil,
			`{"users":[{"id":"a","updatedAt":1},{"id":"b","updatedAt":1}]}`,
			`{"users":[{"id":"a","updatedAt":2},{"id":"b","updatedAt":1}]}`)
		if len(ops) != 0 {
			t.Fatalf("want [], got %s", dumpOps(ops))
		}
	})
	t.Run("real-change-survives-ignored-drift", func(t *testing.T) {
		ops := ignoreDiff(t, pkUsersSchema, []string{"/users/*/updatedAt"}, nil,
			`{"users":[{"id":"a","updatedAt":1,"name":"A"}]}`,
			`{"users":[{"id":"a","updatedAt":2,"name":"A2"}]}`)
		if len(ops) != 1 || ops[0].Path != "/users/0/name" {
			t.Fatalf("want single replace /users/0/name, got %s", dumpOps(ops))
		}
	})
}

func TestIgnorePathsLCS(t *testing.T) {
	t.Run("ignored-field-only", func(t *testing.T) {
		ops := ignoreDiff(t, "", []string{"/items/*/ts"}, nil,
			`{"items":[{"v":1,"ts":1},{"v":2,"ts":1}]}`,
			`{"items":[{"v":1,"ts":9},{"v":2,"ts":1}]}`)
		if len(ops) != 0 {
			t.Fatalf("want [], got %s", dumpOps(ops))
		}
	})
	t.Run("move-pairing-with-ignored-drift", func(t *testing.T) {
		ops := ignoreDiff(t, "", []string{"/items/*/ts"}, []PatcherOption{EmitMoves(true)},
			`{"items":[{"v":"X","ts":1},{"v":"Y","ts":1}]}`,
			`{"items":[{"v":"Y","ts":9},{"v":"X","ts":1}]}`)
		moves, replaces := 0, 0
		for _, o := range ops {
			switch o.Op {
			case OpMove:
				moves++
			case OpReplace:
				replaces++
			}
		}
		if moves != 1 || replaces != 0 {
			t.Fatalf("want 1 move, 0 replace, got %s", dumpOps(ops))
		}
	})
}

func TestIgnorePathsWholesaleDisabled(t *testing.T) {
	schema := `{"type":"object","properties":{"a":{"type":"array","items":{"type":"object","properties":{"t":{"type":"string"},"ts":{"type":"number"}}}}}}`
	ops := ignoreDiff(t, schema, []string{"/a/*/ts"}, []PatcherOption{WholesaleReplaceFallback(true)},
		`{"a":[{"t":"AAAAAAAAAA","ts":1},{"t":"BBBBBBBBBB","ts":2}]}`,
		`{"a":[{"t":"ZZZZZZZZZZ","ts":5},{"t":"YYYYYYYYYY","ts":6}]}`)
	for _, o := range ops {
		if o.Op == OpReplace && o.Path == "/a" {
			t.Fatalf("wholesale replace at /a leaked ignored content: %s", dumpOps(ops))
		}
		if o.Op == OpReplace && strings.HasSuffix(o.Path, "/ts") {
			t.Fatalf("emitted a replace on an ignored /ts field: %s", dumpOps(ops))
		}
	}
}

func TestIgnorePathsByteStable(t *testing.T) {
	// An empty ignore set must be byte-identical to no option (GEN §10.2).
	orig := `{"meta":{"ts":1,"n":"a"},"arr":[{"id":"x","v":1}]}`
	mod := `{"meta":{"ts":2,"n":"b"},"arr":[{"id":"x","v":2}]}`
	base, err := NewPatcher(Plan{})
	if err != nil {
		t.Fatal(err)
	}
	withEmpty, err := NewPatcher(Plan{}, IgnorePaths())
	if err != nil {
		t.Fatal(err)
	}
	o1 := base.Execute(mustDecode(t, orig), mustDecode(t, mod))
	o2 := withEmpty.Execute(mustDecode(t, orig), mustDecode(t, mod))
	if dumpOps(o1) != dumpOps(o2) {
		t.Fatalf("empty ignorePaths not byte-stable:\n base: %s\nempty: %s", dumpOps(o1), dumpOps(o2))
	}
}
